lowranksvd/src/lowranksvd.jl

```diff
@@ -0,0 +1,168 @@
+              choosen according to the following criteria: in general,
+              :math:`k <= l <= min(2*k, m, n)`. For large low-rank
+              matrices, take :math:`l = k + 5..10`.  If k is
+              relatively small compared to :math:`min(m, n)`, choosing
+              :math:`l = k + 0..2` may be sufficient.
+    .. note:: To obtain repeatable results, reset the seed for the
+              pseudorandom number generator
+    Args::
+        A (AbstractArray{T}): the input tensor of size :math:`(m, n)`
+        l (Int64): the dimension of subspace spanned by :math:`Q`
+                 columns.
+        niter (Int64, optional): the number of subspace iterations to
+                               conduct; ``niter`` must be a
+                               nonnegative integer. In most cases, the
+                               default value 2 is more than enough.
+        M (AbstractArray{T}, optional): the input tensor of size
+                              :math:`(m, n)`.
+    References::
+        - Nathan Halko, Per-Gunnar Martinsson, and Joel Tropp, Finding
+          structure with randomness: probabilistic algorithms for
+          constructing approximate matrix decompositions,
+          arXiv:0909.4061 [math.NA; math.PR], 2009 (available at
+          `arXiv <http://arxiv.org/abs/0909.4061>`_).
+    """
+
+    m, n = size(A)
+    Ω = rand(T, (n, l))
+
+    Y = zeros(T, (m, l))
+    Y_H = zeros(T, (n, l))
+
+    if M === nothing 
+        F_j = qr!(mul!(Y, A, Ω))
+        for j = 1:niter
+            F_H_j = qr!(mul!(Y_H, A', Matrix(F_j.Q)))
+            F_j = qr!(mul!(Y, A, Matrix(F_H_j.Q)))
+        end
+    else
+        Z = zeros(T, (m, l))
+        Z_H = zeros(T, (n, l))
+        F_j = qr!(mul!(Y, A, Ω) - mul!(Z, M, Ω))
+        for j = 1:niter
+            F_H_j = qr!(mul!(Y_H, A', Matrix(F_j.Q)) - mul!(Z_H, M', Matrix(F_j.Q)))
+            F_j = qr!(mul!(Y, A, Matrix(F_H_j.Q)) - mul!(Z, M, Matrix(F_H_j.Q)))
+        end
+    end
+    Matrix(F_j.Q)
+end
+
+function _svd_lowrank(A::AbstractArray{T}, l::Int64, niter::Int64 = 2, M::Union{AbstractArray{T}, Nothing} = nothing) where T
+    
+        m, n = size(A)
+    if M === nothing
+        Mt = nothing
+    else
+        Mt = transpose(M)
+    end
+    At = transpose(A)
+
+    if m < n || n > l
+        """
+        computing the SVD approximation of a transpose in
+        order to keep B shape minimal (the m < n case) or the V
+        shape small (the n > l case)
+        """
+        Q = get_approximate_basis(At, l, niter, Mt)
+        Qc = conj(Q)
+        if M === nothing
+            Bt = A * Qc
+        else
+            Bt = A * Qc - M * Qc
+        end
+        U, S, Vt = svd!(Bt)
+        Vt = Vt * Q'
+    else
+        Q = get_approximate_basis(A, l, niter, M)
+        Qc = conj(Q)
+        if M === nothing
+            B = At * Qc
+        else
+            B = At * Qc - Mt * Qc
+        end
+
+        U, S, Vt = svd!(transpose(B))
```

Comment:
不是很确定，不过这样可能会更高效一些：

- 相比于 `transpose(B)`, 直接使用内存连续的`Bt::Matrix` 可能更容易调用高效的 svd subtoutine
- 两个矩阵乘法 => 一个矩阵乘法

```suggestion
        if M === nothing
            Bt = Q' * A
        else
            Bt = Q' * (A - M)
        end

        U, S, Vt = svd!(Bt)
```

Reply:
我很同意老师关于两个矩阵乘法的建议，我参考[pytorch实现](https://github.com/jieli-matrix/svd_lowrank/blob/main/pytorch_implement/get_approximate_basis.py#L35)时也不太理解为什么要拆分，我想到的可能的解释是：1. 可能是想保证A的低秩性结构不被破坏 2. M的mean size为(1,n)可以广播；因为我实现的时候直接指定M与A大小一致，确实如老师所说，合并以后结果一致，FLOPS数也会降低。